javaslang/src/main/java/javaslang/collection/Iterator.java

```diff
@@ -2089,6 +2063,40 @@ public String toString() {
         }
     }
 
+    final class GroupedIterator<T> implements Iterator<Seq<T>> {
+
+        private final Iterator<T> that;
+        private final int step;
+        private final int gap;
+
+        private List<T> buffer;
+
+        GroupedIterator(Iterator<T> that, int size, int step) {
+            if (size < 1 || step < 1) {
+                throw new IllegalArgumentException("size (" + size + ") and step (" + step + ") must both be positive");
+            }
+            this.that = that;
+            this.step = step;
+            this.gap = Math.max(step - size, 0);
+            this.buffer = List.ofAll(that.take(size));
+        }
+
+        @Override
+        public boolean hasNext() {
+            return !buffer.isEmpty();
+        }
+
+        @Override
+        public Seq<T> next() {
+            if (!hasNext()) {
+                throw new NoSuchElementException();
+            }
+            final Seq<T> result = buffer;
+            buffer = that.hasNext() ? that.drop(gap).take(step).toList().prependAll(buffer.drop(step)) : buffer.take(0);
```

Comment:
If you're lucky it will end up on the stack (i.e. escape analysis), you should measure the impact :)
Seems more readable/maintainable though, that's more important for me personally.